pkg/cloudevents/transport/x/example_implementing_test.go

```diff
@@ -0,0 +1,70 @@
+package x_test
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/transport"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/transport/x"
+)
+
+// ExMessage is a json.RawMessage, which is just a byte slice
+// containing a JSON encoded event.
+type ExMessage struct{ json.RawMessage }
+
+func (m ExMessage) Structured() (string, []byte) {
+	return cloudevents.ApplicationCloudEventsJSON, []byte(m.RawMessage)
+}
+
+func (m ExMessage) Event() (e cloudevents.Event, err error) {
+	err = json.Unmarshal(m.RawMessage, &e)
+	return e, err
+}
+
+func (m ExMessage) Finish(error) {}
+
+// ExSender sends by writing JSON encoded events to an io.Writer
+type ExSender struct{ *json.Encoder }
+
+func NewExSender(w io.Writer) ExSender { return ExSender{json.NewEncoder(w)} }
+
+func (s ExSender) Send(ctx context.Context, m x.Message) error {
+	if t, b := m.Structured(); t != "" {
+		// Fast case: if the Message is already structured JSON we can
+		// send it directly, no need to decode and re-encode. Encoding a
+		// json.RawMessage to a json.Encoder() is just a byte-buffer copy.
+		return s.Encode(json.RawMessage(b))
```

Comment:
there is no impl for `Encode`. I am confused why you would need to encode message. In the current SDK, message is created by passing an event through a codec.

Reply:
Encode here is the standard json.Encoder.Encode() - the naming is confusing. 

The new Message is an abstraction that can be passed *between* transports, so a Sender may be sending a Message that was encoded by a foreign Receiver. Message advertises if it already contains a structured encoding that the sender can potentially re-use with no re-encoding in the optimized case, but otherwise the sender will call Event() to extract (and possibly decode using some foreign codec)  the Event and re-encode it using its own Codec. I didn't implement an explicit Codec in the example, I just used json calls directly - I probably should for clarity.

Codec still has  role in creating the internal message format for each transport, and may be used directly by users that just want to create e.g. a HTTP message and then use their own client or server code. However for communicating from a Receiver to a Sender, the relevant codecs are used internally by each one, and the Message provides a uniform way to decode the message without knowing what Codec was used to create it.

Reply:
ah, ok so Message has to have some state in it to define in which transport the message is encoded so it can select the correct codec to convert it out. 

It seems like a really interesting feature to be able to move the Message between transports. Middleware will be much faster this way too!